### PR TITLE
Tx generator, beta-version.

### PR DESCRIPTION
**Pull request by denisshevchenko**
This is beta-version of the new tx generator. It's not finished yet, and still contain some hard-coded values (for development simplicity), but it must be a part of `master`, to avoid endless rebasing to the `master` because of massive changes.

### CODE REVIEW

File: cardano-node/app/cardano-cli.hs [hunk 1]
```diff
@@ -25,20 +25,30 @@ import           Cardano.Crypto ( AProtocolMagic(..)
                                 , ProtocolMagicId(..)
                                 , RequiresNetworkMagic(..)
                                 , decodeHash)
-
```
Comment (deepfire) [inline cardano-node/app/cardano-cli.hs]: This separator between local imports and those from other repos should stay.

File: cardano-node/app/cardano-cli.hs [hunk 2]
```diff
@@ -25,20 +25,30 @@ import           Cardano.Crypto ( AProtocolMagic(..)
...
 import           Cardano.CLI.Genesis
 import           Cardano.CLI.Key
+import           Cardano.Config.Logging (LoggingCLIArguments (..),
+                                         createLoggingFeature)
 import           Cardano.CLI.Ops (decideCLIOps)
 import           Cardano.CLI.Run
 
 main :: IO ()
 main = do
-  co <- Opt.customExecParser pref opts
```
Comment (deepfire) [inline cardano-node/app/cardano-cli.hs]: Why did you remove customExecParser?

File: cardano-node/app/cardano-cli.hs [hunk 3]
```diff
@@ -25,20 +25,30 @@ import           Cardano.Crypto ( AProtocolMagic(..)
...
+  co <- Opt.execParser opts
   ops <- decideCLIOps (protocol co)
-  cmdRes <- runExceptT . runCommand ops $ mainCommand co
+  -- Initialize logging layer. Particularly, we need it for benchmarking (command 'generate-txs').
+  let cardanoConfiguration = mainnetConfiguration
+  let cardanoEnvironment   = NoEnvironment
+  finalConfig <- mkConfiguration cardanoConfiguration (commonCli co)
+  (loggingLayer, _loggingFeature) <- createLoggingFeature cardanoEnvironment finalConfig (loggingCli co)
+
+  cmdRes <- runExceptT $ runCommand ops (mainCommand co) loggingLayer
```
Comment (deepfire) [inline cardano-node/app/cardano-cli.hs]: Why those changes?

File: cardano-node/app/cardano-cli.hs [hunk 4]
```diff
@@ -60,17 +67,20 @@ opts =
...
-  = CLI
-      { protocol    :: Protocol,
-        mainCommand :: ClientCommand
-        }
+data CLI = CLI
+  { protocol    :: Protocol
+  , mainCommand :: ClientCommand
+  , commonCli   :: CommonCLI
+  , loggingCli  :: LoggingCLIArguments
+  }
```
Comment (deepfire) [inline cardano-node/app/cardano-cli.hs]: Why this change of CLI?

File: cardano-node/src/Cardano/CLI/Run.hs [hunk 5]
```diff
@@ -53,17 +61,29 @@ import qualified Cardano.Crypto.Hashing as Crypto
...
 import           Cardano.Common.Protocol
-import           Cardano.Node.Configuration.Topology
+import           Cardano.Config.CommonCLI
+import           Cardano.Config.Logging (LoggingLayer (..), Trace)
+import           Cardano.Config.Presets (mainnetConfiguration)
+import           Cardano.Node.Configuration.Topology (TopologyInfo)
 
+import           Ouroboros.Consensus.Ledger.Byron.Config (ByronConfig)
+import           Ouroboros.Consensus.Ledger.Byron (ByronBlockOrEBB (..), GenTxId) 
+import qualified Ouroboros.Consensus.Protocol as Consensus
```
Comment (deepfire) [inline cardano-node/src/Cardano/CLI/Run.hs]: These external imports should come before local imports.

File: cardano-node/app/cardano-cli.hs (hunk 4 above)
Comment (deepfire) [inline cardano-node/app/cardano-cli.hs]: After some thinking, this is a good idea, to avoid repeating the same init across different commands.

File: cardano-node/src/Cardano/CLI/Run.hs [hunk 6]
```diff
@@ -137,78 +157,139 @@ data ClientCommand
...
-runCommand co (SpendUTxO (NewTxFile ctTx) ctKey ins outs common) = do
+runCommand co (SpendUTxO (NewTxFile ctTx) ctKey ins outs common) _ = do
   cc <- liftIO $ mkConfiguration mainnetConfiguration common
   sk <- readSigningKey co ctKey
   gTx <- liftIO $ issueUTxOExpenditure co ins outs cc sk
   liftIO . ensureNewFileLBS ctTx $ serialise gTx
 
+runCommand co
+           (GenerateTxs topology numOfTxs numOfOutsPerTx feePerTx tps sigKeysFiles common)
+           loggingLayer = do
```
Comment (deepfire) [inline cardano-node/src/Cardano/CLI/Run.hs]: This should be a short case branch that calls out to the main function.

File: cardano-node/src/Cardano/CLI/Run.hs [hunk 7]
```diff
@@ -137,78 +157,139 @@ data ClientCommand
...
+  cc <- liftIO $ mkConfiguration mainnetConfiguration common
+  SomeProtocol p <- liftIO $ fromProtocol cc (coProtocol co)
+  case p of
+    protocol@(Consensus.ProtocolRealPBFT _ _ _ _ _) -> do
+      let basicTr :: Trace IO Text
```
Comment (deepfire) [inline cardano-node/src/Cardano/CLI/Run.hs]: We should replace this `let` with a `where` -- so that it's immediately clear what is the action it performs.

The long chain of `let`'s is really obscuring the big picture here.

File: cardano-node/src/Cardano/CLI/Tx/BenchmarkingNodeToNode.hs [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+import           Network.Socket (AddrInfo, SockAddr)
+import           Network.TypedProtocol.Driver (TraceSendRecv, runPeer)
+import           Network.TypedProtocol.Driver.ByteLimit (DecoderFailureOrTooMuchInput)
+import           Ouroboros.Consensus.Block (BlockProtocol)
+import           Ouroboros.Consensus.Mempool.API (GenTxId, GenTx)
+import           Ouroboros.Consensus.Node.Run (RunNode)
+import           Ouroboros.Consensus.NodeNetwork (ProtocolCodecs(..), protocolCodecs)
+import           Ouroboros.Consensus.Protocol.Abstract (NodeConfig)
+import           Ouroboros.Network.Mux (OuroborosApplication(..))
+import           Ouroboros.Network.NodeToNode
```
Comment (deepfire) [inline cardano-node/src/Cardano/CLI/Tx/BenchmarkingNodeToNode.hs]: We avoid use of implicit imports -- this should be wither qualified (`as Network`, for example) or explicit.

File: cardano-node/src/Cardano/CLI/Tx/BenchmarkingTxSubClient.hs [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+{-# LANGUAGE RankNTypes          #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Cardano.CLI.Tx.BenchmarkingTxSubClient
+  ( txSubmissionClient
+  ) where
+
+import Prelude
+
+import Control.Monad.Class.MonadSTM
```
Comment (deepfire) [inline cardano-node/src/Cardano/CLI/Tx/BenchmarkingTxSubClient.hs]: Should be an explicit import, I suppose..

File: cardano-node/src/Cardano/CLI/Tx/BenchmarkingTxSubmission.hs [hunk 10]
```diff
@@ -0,0 +1,350 @@
...
+  ( ROEnv(..)
+  , RPCTxSubmission(..)
+  , TraceBenchTxSubmit(..)
+  , bulkSubmission
+  )
+where
+
+import           Prelude
+import           Cardano.Prelude (MonadIO, Word16, Seq, StateT
+                                 , evalStateT, gets, isJust, lift, {- liftIO, -} modify, toList, unless, when)
```
Comment (deepfire) [inline cardano-node/src/Cardano/CLI/Tx/BenchmarkingTxSubmission.hs]: Preludes are exempt from the explicit import list requirement -- so this one can be removed.

File: cardano-node/src/Cardano/CLI/Tx/BenchmarkingTxSubmission.hs [hunk 11]
```diff
@@ -0,0 +1,350 @@
...
+--   configurable latency hiding (number of transactions to try to
+--   keep ahead of actual submission) and has tracing that both
+--   reports observables related to transaction submission and whether
+--   the tx generation is able to keep up the pace.
+--
+--   For the details of the protocol description see
+--   `TxSubmit.TxSubmission` and its associated instances
+--   (e.g. `Message`).
+
+bulkSubmission :: forall m blk txid tx .
```
Comment (deepfire) [inline cardano-node/src/Cardano/CLI/Tx/BenchmarkingTxSubmission.hs]: It's better to move the `:: forall m blk txid tx .` to the next line and reindent -- so that the offset of the entire block doesn't depend on the length of the `bulkSubmission` name.

File: cardano-node/src/Cardano/CLI/Tx/Submission.hs [hunk 12]
```diff
@@ -9,6 +9,9 @@
 
 module Cardano.CLI.Tx.Submission (
       handleTxSubmission
+    , localSocketFilePath
+    , localSocketAddrInfo
```
Comment (deepfire) [inline cardano-node/src/Cardano/CLI/Tx/Submission.hs]: We maintain alphabetic sort order in the export list.

File: cardano-node/src/Cardano/CLI/Tx/BenchmarkingTxSubmission.hs (hunk 10 above)
Comment (denisshevchenko) [reply cardano-node/src/Cardano/CLI/Tx/BenchmarkingTxSubmission.hs]: Neil made it to avoid conflicts with `Prelude`.

File: cardano-node/src/Cardano/CLI/Run.hs (hunk 6 above)
Comment (denisshevchenko) [reply cardano-node/src/Cardano/CLI/Run.hs]: Sorry, didn't understand it...

File: cardano-node/src/Cardano/CLI/Run.hs [hunk 13]
```diff
@@ -30,14 +30,22 @@ module Cardano.CLI.Run (
...
   , TxFile(..)
   , NewTxFile(..)
+  , NumberOfTxs(..)
+  , NumberOfOutputsPerTx(..)
+  , FeePerTx(..)
+  , TPSRate(..)
   ) where
 
-import           Cardano.Prelude hiding (option, show, trace)
+import Data.String(String)
```
Comment (deepfire) [inline cardano-node/src/Cardano/CLI/Run.hs]: Here we currently mandate vertical alignment.

File: cardano-node/src/Cardano/CLI/Run.hs [hunk 14]
```diff
@@ -48,22 +56,37 @@ import qualified Cardano.Chain.Delegation as Delegation
 import qualified Cardano.Chain.Genesis as Genesis
 import           Cardano.Chain.Slotting (EpochNumber(..))
 import qualified Cardano.Chain.UTxO as UTxO
+
```
Comment (deepfire) [inline cardano-node/src/Cardano/CLI/Run.hs]: I wouldn't say the intra-external-module import white space is necessary, but it's not critical.

File: cardano-node/src/Cardano/CLI/Run.hs (hunk 6 above)
Comment (deepfire) [reply cardano-node/src/Cardano/CLI/Run.hs]: Because we don't want functions that are too long.

`runCommand` went through refactoring to avoid precisely that.

File: cardano-node/src/Cardano/CLI/Tx/Generation.hs [hunk 15]
```diff
@@ -0,0 +1,850 @@
...
+  -> Maybe CC.Common.Address              -- the address to associate with the 'change',
+                                          -- if different from that of the first argument
+  -> Set (r, CC.UTxO.TxOut)               -- each recipient and their payment details
+  -> ( Maybe (Word32, CC.Common.Lovelace) -- the 'change' index and value (if any)
+     , CC.Common.Lovelace                 -- the associated fees
+     , Map r Word32                       -- the offset map in the transaction below
+     , GenTx (ByronBlockOrEBB cfg)
+     )
+mkTransaction cfg ((txinFrom, txoutFrom), signingKey) mChangeAddress payments =
+  let paymentsList = toList payments
```
Comment (deepfire) [inline cardano-node/src/Cardano/CLI/Tx/Generation.hs]: I think this would also benefit if rewritten with a `where` clause instead of a `let`.

File: cardano-node/src/Cardano/CLI/Tx/Generation.hs [hunk 16]
```diff
@@ -0,0 +1,850 @@
...
+  -> Tracer IO String
+  -> CardanoConfiguration
+  -> Consensus.Protocol (ByronBlockOrEBB ByronConfig)
+  -> TopologyInfo
+  -> NumberOfTxs
+  -> NumberOfOutputsPerTx
+  -> FeePerTx
+  -> TPSRate
+  -> [FilePath]
+  -> IO ()
```
Comment (deepfire) [inline cardano-node/src/Cardano/CLI/Tx/Generation.hs]: I think many of the arguments should be grouped into a `data` type, like `BenchmarkConfig` or somesuch.
Same for the tracers, they should also benefit from grouping -- `data BencmarkTracers` maybe..

This would help with being able to just pass it down to `runBenchmark`.

File: cardano-node/src/Cardano/CLI/Tx/Generation.hs [hunk 17]
```diff
@@ -0,0 +1,850 @@
...
+  -> NodeConfig ByronEBBExtNodeConfig
+  -> CC.Common.Address
+  -> Crypto.SigningKey
+  -> CC.Common.Address
+  -> TopologyInfo
+  -> NumberOfTxs
+  -> NumberOfOutputsPerTx
+  -> FeePerTx
+  -> TPSRate
+  -> IO ()
```
Comment (deepfire) [inline cardano-node/src/Cardano/CLI/Tx/Generation.hs]: Oh, yes, a dedicated `data` type would help here..

File: cardano-node/src/Cardano/CLI/Tx/Generation.hs [hunk 18]
```diff
@@ -0,0 +1,850 @@
...
+        :: ROEnv (Byron.GenTxId (ByronBlockOrEBB ByronConfig)) (GenTx (ByronBlockOrEBB ByronConfig))
+        -> ROEnv (Byron.GenTxId (ByronBlockOrEBB ByronConfig)) (GenTx (ByronBlockOrEBB ByronConfig))
+      updROEnv defaultROEnv =
+        ROEnv { targetBacklog     = targetBacklog defaultROEnv
+              , txNumServiceTime  = Nothing
+              , txSizeServiceTime = Nothing
+              }
+
+  -- Run generator.
+  txGenerator tracer
```
Comment (deepfire) [inline cardano-node/src/Cardano/CLI/Tx/Generation.hs]: Maybe a `data` type for `txGenerator` as well? `TxGeneratorParams` or somesuch..

File: cardano-node/src/Cardano/CLI/Tx/BenchmarkingTxSubmission.hs [hunk 19]
```diff
@@ -0,0 +1,353 @@
...
+    -- Update terminating, if needed 
+    -- lift . traceWith tr . TraceBenchTxSubDebug $ "go1, update terminating, if needed"
+    when (nowTerminating && not terminating') $ do
+      -- lift . traceWith tr . TraceBenchTxSubDebug $ "go1, update terminating: True"
+      modify (\x -> x {terminating = True})
+
+    -- incorporate new transactions from the generator into
+    -- NotYetSent
+    -- lift . traceWith tr . TraceBenchTxSubDebug $ "go1, incorporate new transactions from the generator into NotYetSent"
+    flip (maybe (pure ())) newTxs $ \xs -> do
```
Comment (deepfire) [inline cardano-node/src/Cardano/CLI/Tx/BenchmarkingTxSubmission.hs]: Hmm.. is it just me, or is this somewhat hard to read? 

Do you think we could have a simpler presentation -- maybe a dedicated function?

File: cardano-node/src/Cardano/CLI/Tx/BenchmarkingTxSubmission.hs [hunk 20]
```diff
@@ -0,0 +1,353 @@
...
+          reqTxIds <- takeTMVar rpcIn -- no other message should occur as nothing is in flight
+          case reqTxIds of
+            RPCRequestTxIds _ resp -> putTMVar resp Nothing
+            _ -> return ()
+    else do -- process next interaction
+      -- lift . traceWith tr . TraceBenchTxSubDebug $ "go, process next interaction"
+      go1 env
+
+  go1 :: ROEnv txid tx -> StateT (RWEnv m txid tx) m ()
+  go1 env = do
```
Comment (deepfire) [inline cardano-node/src/Cardano/CLI/Tx/BenchmarkingTxSubmission.hs]: This function feels like it could greatly benefit from being split into parts, at least as a pipeline.

File: cardano-node/src/Cardano/CLI/Run.hs (hunk 6 above)
Comment (denisshevchenko) [reply cardano-node/src/Cardano/CLI/Run.hs]: Done. Now I just pass `loggingLayer` in my function, all tracers will be created on my side.

File: cardano-node/src/Cardano/CLI/Tx/BenchmarkingTxSubmission.hs (hunk 19 above)
Comment (denisshevchenko) [reply cardano-node/src/Cardano/CLI/Tx/BenchmarkingTxSubmission.hs]: This code was written by Neil. To be honest, I don't want to touch it right now (actually, only Neil really knows how this code works).

File: cardano-node/src/Cardano/CLI/Tx/BenchmarkingTxSubmission.hs (hunk 20 above)
Comment (denisshevchenko) [reply cardano-node/src/Cardano/CLI/Tx/BenchmarkingTxSubmission.hs]: This code was written by Neil. To be honest, I don't want to touch it right now (actually, only Neil really knows how this code works).

File: cardano-node/src/Cardano/CLI/Run.hs [hunk 21]
```diff
@@ -137,78 +148,107 @@ data ClientCommand
...
   liftIO . ensureNewFileLBS ctTx $ serialise gTx
 
+runCommand co
+           (GenerateTxs topology numOfTxs numOfOutsPerTx feePerTx tps sigKeysFiles common)
+           loggingLayer = do
+  cc <- liftIO $ mkConfiguration mainnetConfiguration common
+  SomeProtocol p <- liftIO $ fromProtocol cc (coProtocol co)
+  case p of
+    protocol@(Consensus.ProtocolRealPBFT _ _ _ _ _) ->
+      liftIO $ genesisBenchmarkRunner loggingLayer
```
Comment (deepfire) [inline cardano-node/src/Cardano/CLI/Run.hs]: Here you can use `withRealPBFT`.

Comment (denisshevchenko) [reply cardano-node/src/Cardano/CLI/Run.hs]: Done.

**Comment by deepfire** [comment]
I've asked @intricate to have an independent go at this PR.

File: cardano-node/src/Cardano/CLI/Tx/Generation.hs [hunk 22]
```diff
@@ -0,0 +1,893 @@
...
+                    pInfoConfig
+                    topologyInfo
+                    genesisUtxo
+                    genesisAddress
+                    targetAddress = do
+  let (mFunds, _, _initGenTx) = extractInitialFunds pInfoConfig
+                                                    (genesisUtxo Map.! 0) -- corresponds to 'genesisAddress'.
+                                                    targetAddress
+
+  let ((_, _), signingKey) = genesisUtxo Map.! 0
```
Comment (intricate) [inline cardano-node/src/Cardano/CLI/Tx/Generation.hs]: I wonder if it'd be nicer to utilize `Map.!?`, pattern match on the result, and then explicitly call `error` if it's `Nothing`. This way we could also provide a more helpful error message detailing the module and function where the error occurred (i.e. `Cardano.CLI.Tx.Generation.prepareInitialFunds`). 

Also, it seems like we utilize `genesisUtxo Map.! 0` twice in this function so we could probably give it its own `let` binding.

@deepfire: think I'm nitpicking too much?

File: cardano-node/src/Cardano/CLI/Tx/Generation.hs [hunk 23]
```diff
@@ -0,0 +1,893 @@
...
+               remoteAddr
+               updROEnv
+               txsForSubmission
+
+-- | At this moment 'sourceAddress' contains a huge amount of money (lets call it A).
+--   Now we have to split this amount to N equal parts, as a result we'll have
+--   N UTxO entries, and alltogether these entries will contain the same amount A.
+--   E.g. (1 entry * 1000 ADA) -> (10 entries * 100 ADA).
+--   Technically all splitting transactions will send money back to 'sourceAddress'.
+createMoreFundCoins
```
Comment (intricate) [inline cardano-node/src/Cardano/CLI/Tx/Generation.hs]: Nice comments btw :+1:. Appreciate you documenting much of what is going on.

File: cardano-node/src/Cardano/CLI/Tx/Generation.hs [hunk 24]
```diff
@@ -0,0 +1,893 @@
...
+                                          -- input and the key to spend the
+                                          -- associated value
+  -> Maybe CC.Common.Address              -- the address to associate with the 'change',
+                                          -- if different from that of the first argument
+  -> Set (r, CC.UTxO.TxOut)               -- each recipient and their payment details
+  -> ( Maybe (Word32, CC.Common.Lovelace) -- the 'change' index and value (if any)
+     , CC.Common.Lovelace                 -- the associated fees
+     , Map r Word32                       -- the offset map in the transaction below
+     , GenTx (ByronBlockOrEBB cfg)
+     )
```
Comment (intricate) [inline cardano-node/src/Cardano/CLI/Tx/Generation.hs]: Not a big deal but, typically, once I start dealing with anything more than a triple, I like to define a separate data type.

File: cardano-node/src/Cardano/CLI/Tx/BenchmarkingTxSubmission.hs [hunk 25]
```diff
@@ -0,0 +1,353 @@
...
+  --   peer.
+  | TraceBenchTxSubServAck [txid]
+  -- ^ An ack (window moved over) received for these transactions.
+  | TraceBenchTxSubIdle
+  -- ^ Remote peer requested new transasctions but none were
+  --   available, generator not keeping up?
+  | TraceBenchTxSubRateLimit DiffTime
+  -- ^ Rate limiter bit, this much delay inserted to keep within
+  --   configured rate.
+  | TraceBenchTxSubDebug String
```
Comment (intricate) [inline cardano-node/src/Cardano/CLI/Tx/BenchmarkingTxSubmission.hs]: Can we doc this constructor too?

File: cardano-node/src/Cardano/CLI/Tx/BenchmarkingTxSubmission.hs [hunk 26]
```diff
@@ -0,0 +1,353 @@
...
+  , txSizeServiceTime = Nothing
+  }
+
+data RWEnv m txid tx = RWEnv
+  { terminating     :: Bool
+  , activityState   :: ActivityState
+  , proceedAfter    :: Time m
+  , availableOp     :: Maybe (Int, LazyTMVar m (Maybe [(txid, TxSubmit.TxSizeInBytes)]))
+  -- ^ the window and the response action
+  , inFlight
```
Comment (intricate) [inline cardano-node/src/Cardano/CLI/Tx/BenchmarkingTxSubmission.hs]: What's going on with this `inFlight` field? Is this valid code?

**Comment by intricate** [comment]
> It has been agreed with @denisshevchenko that this PR is merged as a WIP, to reduce the rebase pain, and that further improvements will be made.

Hopefully, going forward, we can try to avoid raising large PRs like this in the future. It's quite difficult to actually give it a solid review and then there's the aforementioned rebase pain.

File: cardano-node/src/Cardano/CLI/Tx/BenchmarkingTxSubmission.hs (hunk 26 above)
Comment (intricate) [reply cardano-node/src/Cardano/CLI/Tx/BenchmarkingTxSubmission.hs]: The CI checks are all passing so I'm guessing it's valid, but I'm wondering why there's no type signature?

**Comment by denisshevchenko** [comment]
> Hopefully we can try to avoid raising large PRs like this in the future. It's quite difficult to actually give it a solid review and then there's the aforementioned rebase pain.

@intricate Agree. Tx generator is a quite big piece of code, but after this branch will be merged in the `master`, following PRs will be much, much smaller.

File: cardano-node/src/Cardano/CLI/Tx/BenchmarkingTxSubmission.hs (hunk 25 above)
Comment (denisshevchenko) [reply cardano-node/src/Cardano/CLI/Tx/BenchmarkingTxSubmission.hs]: I'm not sure if we need this constructor at all. It was used as a fastest way to get raw debug messages..

File: cardano-node/src/Cardano/CLI/Tx/Generation.hs [hunk 27]
```diff
@@ -0,0 +1,893 @@
...
+
+launchTxPeer
+  :: forall m block txid tx.
+     ( RunNode block
+     , m ~ IO
+     , txid ~ Byron.GenTxId block
+     , tx ~ GenTx block
+     )
+  => Tracer m (TraceBenchTxSubmit txid)
+  -- Tracer carrying the benchmarking events
```
[the remark this answers is not in the document]
Comment (denisshevchenko) [reply cardano-node/src/Cardano/CLI/Tx/Generation.hs]: I did it because parsing problems from Haddock. I don't know why...

File: cardano-node/src/Cardano/CLI/Tx/BenchmarkingTxSubClient.hs [hunk 28]
```diff
@@ -0,0 +1,59 @@
...
+                                                                 TxSubmissionClient(..))
+import           Ouroboros.Network.Protocol.TxSubmission.Type (BlockingReplyList(..),
+                                                               TokBlockingStyle(..))
+
+txSubmissionClient
+  :: forall m block txid tx .
+     ( MonadSTM m
+     , txid ~ GenTxId block
+     , tx ~ GenTx block)
+  => LazyTMVar m (RPCTxSubmission m txid tx)
```
[the remark this answers is not in the document]
Comment (denisshevchenko) [reply cardano-node/src/Cardano/CLI/Tx/BenchmarkingTxSubClient.hs]: > Do we specifically need a LazyTMVar here 

Honestly, I don't know (this code was written by Neil).

File: cardano-node/src/Cardano/CLI/Tx/Generation.hs (hunk 22 above)
Comment (denisshevchenko) [reply cardano-node/src/Cardano/CLI/Tx/Generation.hs]: > I wonder if it'd be nicer to utilize `Map.!?`, pattern match on the result, and then explicitly call `error` if it's `Nothing`.

It cannot be `Nothing` here, because of `extractGenesisFunds` function (which already called).

File: cardano-node/src/Cardano/CLI/Tx/BenchmarkingTxSubmission.hs (hunk 26 above)
Comment (deepfire) [reply cardano-node/src/Cardano/CLI/Tx/BenchmarkingTxSubmission.hs]: Wow.
